### PR TITLE
Several StaffBot improvements

### PR DESCRIPTION
**Pull request by kkamalov**

### CODE REVIEW

File: orchestra/bots/assignment_policies.py [hunk 1]
```diff
@@ -1,9 +1,20 @@
...
 
 
 def staffbot_autoassign(task, **kwargs):
     request_cause = StaffBotRequest.RequestCause.AUTOSTAFF.value
     bot = StaffBot()
     bot.staff(task.id, request_cause=request_cause)
     return task
+
+
+def staffbot_request(task, **kwargs):
```
Comment (marcua) [inline orchestra/bots/assignment_policies.py]: what's this for?

File: orchestra/communication/staffing.py [hunk 2]
```diff
@@ -24,6 +28,12 @@
 from orchestra.utils.task_lifecycle import is_worker_certified_for_task
 
 
+MAX_REQUEST_COUNT = 3
+MAX_TIME_NO_RESPONSE = 300  # Minutes or 5 hours
+RESEND_TIME = 60  # Minutes or 1 hours
+STAFFING_GROUP_ID = 'G7R9Q24TZ'
```
Comment (marcua) [inline orchestra/communication/staffing.py]: all of these should be in `SETTINGS`

File: orchestra/communication/staffing.py [hunk 3]
```diff
@@ -197,3 +227,56 @@ def get_available_requests(worker):
...
+        Task.objects.all()
+        .filter(staffing_requests__inquiries__responses__is_winner=False)
+        .exclude(staffing_requests__isnull=True)
+        .exclude(staffing_requests__inquiries__isnull=True)
+        .order_by('-start_datetime')
+        .values('staffing_requests__required_role_counter', 'id')
+        .annotate(Count('staffing_requests__id')))
+
+    for task_value in task_values:
+        if task_value['staffing_requests__id__count'] > MAX_REQUEST_COUNT:
```
Comment (marcua) [inline orchestra/communication/staffing.py]: i probably don't understand this one, but if you have MAX_TIME_NO_RESPONSE, i'd delete this thing---essentially, we shouldn't send both messages, so just delete this implementation and keep the time-based one.

File: orchestra/communication/staffing.py [hunk 4]
```diff
@@ -24,6 +28,12 @@
 from orchestra.utils.task_lifecycle import is_worker_certified_for_task
```
Comment (marcua) [inline orchestra/communication/staffing.py]: delete: MAX_REQUEST_COUNT
delete: RESEND_TIME (unused)
rename: ORCHESTRA_STAFFBOT_STAFFING_MAX_TIME
 -> make this a timedelta, get rid of comment :)
rename: ORCHESTRA_STAFFBOT_ STAFFING_GROUP_ID

File: orchestra/templates/communication/tasks_available_email.txt [hunk 5]
```diff
@@ -0,0 +1,5 @@
+Hello {{user.first_name}}!
+
+Tasks are available for you to work on, if you'd like!  Go through a link to for details:
```
Comment (marcua) [inline orchestra/templates/communication/tasks_available_email.txt]: Tasks are still available for you to work on, if you'd like!  Here are some more details:

File: orchestra/templates/communication/tasks_available_email.txt [hunk 6]
```diff
@@ -0,0 +1,5 @@
...
+
+<a href="{{tasks_url}}> View Tasks</a>
```
Comment (marcua) [inline orchestra/templates/communication/tasks_available_email.txt]: no space before `View`

File: orchestra/templates/communication/tasks_available_slack.txt [hunk 7]
```diff
@@ -0,0 +1,5 @@
+Hello {{user.first_name}}!
+
+Tasks are available for you to work on, if you'd like!  Go through a link to for details:
```
Comment (marcua) [inline orchestra/templates/communication/tasks_available_slack.txt]: Tasks are still available for you to work on, if you'd like!  Here are some more details:

File: orchestra/utils/task_lifecycle.py [hunk 8]
```diff
@@ -37,6 +41,40 @@ class AssignmentPolicyType(object):
     REVIEWER = 'reviewer'
 
 
+def mark_worker_as_winner(staffbot_request, worker):
```
Comment (marcua) [inline orchestra/utils/task_lifecycle.py]: this should live in the staffing code, perhaps in `orchestra/communication/staffing.py`

File: orchestra/utils/task_lifecycle.py [hunk 9]
```diff
@@ -37,6 +41,40 @@ class AssignmentPolicyType(object):
     REVIEWER = 'reviewer'
```
Comment (marcua) [inline orchestra/utils/task_lifecycle.py]: this duplicates some code in orchestra/communication/staffing.py:handle_staffing_response, which also calls `assign_task`. 

maybe make a `mark_worker_as_winner` get called by `assign_task` and `reassign_task` and then don't declare the winner int he `handle_staffing_response` function, but pass a default communication preference or staffing inquiry to use?

File: orchestra/utils/task_lifecycle.py [hunk 10]
```diff
@@ -237,6 +276,12 @@ def assign_task(worker_id, task_id):
         assignment=assignment,
         start_datetime=assignment.start_datetime)
 
+    staffbot_request = StaffBotRequest.objects.filter(
```
Comment (marcua) [inline orchestra/utils/task_lifecycle.py]: also in `reassign_task`

File: orchestra/communication/staffing.py [hunk 11]
```diff
@@ -197,3 +227,56 @@ def get_available_requests(worker):
             reverse('orchestra:communication:available_staffing_requests'))
         contexts.append(metadata)
     return contexts
+
+
+def get_inquiries_per_worker_count(request):
```
Comment (marcua) [inline orchestra/communication/staffing.py]: unused/delete?

File: orchestra/communication/staffing.py [hunk 12]
```diff
@@ -197,3 +227,56 @@ def get_available_requests(worker):
...
+            continue
+
+
+def available_tasks_followup():
+    staffbot = StaffBot()
+    workers = Worker.objects.all()
+    for worker in workers:
+        requests = get_available_requests(worker)
+        if len(requests):
+            # send staffbot request.
```
Comment (marcua) [inline orchestra/communication/staffing.py]: can delete comment :)

File: orchestra/communication/staffing.py [hunk 13]
```diff
@@ -197,3 +227,56 @@ def get_available_requests(worker):
...
+                 .format(request.task, request.created_at)))
+            continue
+
+
+def available_tasks_followup():
+    staffbot = StaffBot()
+    workers = Worker.objects.all()
+    for worker in workers:
+        requests = get_available_requests(worker)
+        if len(requests):
```
Comment (marcua) [inline orchestra/communication/staffing.py]: if easy, make this happen if all of the requests are older than than ORCHESTRA_STAFFBOT_MIN_FOLLOWUP_TIME

File: orchestra/communication/staffing.py [hunk 14]
```diff
@@ -197,3 +227,56 @@ def get_available_requests(worker):
...
+            continue
+        required_role_counter = task_value[
+            'staffing_requests__required_role_counter']
+        request = (
+            StaffBotRequest.objects.filter(
+                task=task_value['id'],
+                required_role_counter=required_role_counter)
+            .order_by('-created_at'))[0]
+
+        if request.created_at < timezone.now() - timedelta(
```
Comment (marcua) [inline orchestra/communication/staffing.py]: if you delete `MAX_REQUEST_COUNT` logic, you can push this request.created_at into the task_values filter :)

File: orchestra/communication/staffing.py [hunk 15]
```diff
@@ -197,3 +227,56 @@ def get_available_requests(worker):
...
+                minutes=MAX_TIME_NO_RESPONSE):
+            message_experts_slack_group(
+                STAFFING_GROUP_ID,
+                ('No winner request for task {}! Created at {}'
+                 .format(request.task, request.created_at)))
+            continue
+
+
+def available_tasks_followup():
```
Comment (marcua) [inline orchestra/communication/staffing.py]: do you also want to call `check_unstaffed_tasks` in here?

Comment (marcua) [inline orchestra/communication/staffing.py]: add test

File: orchestra/communication/staffing.py [hunk 16]
```diff
@@ -197,3 +227,56 @@ def get_available_requests(worker):
...
+
+
+def get_inquiries_per_worker_count(request):
+    inq_count = (request.inquiries.all()
+                 .values('communication_preference__worker')
+                 .annotate(Count('communication_preference__worker')))
+    return inq_count[0]['communication_preference__worker__count']
+
+
+def check_unstaffed_tasks():
```
Comment (marcua) [inline orchestra/communication/staffing.py]: add test

**Review by marcua — CHANGES_REQUESTED** [review]
Yay! I can't wait to see this get in!

File: orchestra/communication/staffing.py (hunk 15 above)
Comment (marcua) [inline orchestra/communication/staffing.py]: `remind_workers_about_available_tasks`?

File: orchestra/communication/staffing.py (hunk 16 above)
Comment (marcua) [inline orchestra/communication/staffing.py]: `warn_staffing_team_about_unstaffed_tasks`?

File: orchestra/utils/task_lifecycle.py [hunk 17]
```diff
@@ -16,8 +16,12 @@
 from orchestra.core.errors import TaskDependencyError
```
Comment (marcua) [inline orchestra/utils/task_lifecycle.py]: might have to update tests for these code changes :)

File: orchestra/bots/assignment_policies.py (hunk 1 above)
Comment (kkamalov) [reply orchestra/bots/assignment_policies.py]: for updated workflow to create staffbot requests

File: orchestra/communication/staffing.py (hunk 4 above)
Comment (kkamalov) [reply orchestra/communication/staffing.py]: done

File: orchestra/communication/staffing.py (hunk 2 above)
Comment (kkamalov) [reply orchestra/communication/staffing.py]: done

File: orchestra/communication/staffing.py (hunk 11 above)
Comment (kkamalov) [reply orchestra/communication/staffing.py]: done

File: orchestra/communication/staffing.py (hunk 16 above)
Comment (kkamalov) [reply orchestra/communication/staffing.py]: done

File: orchestra/communication/staffing.py (hunk 15 above)
Comment (kkamalov) [reply orchestra/communication/staffing.py]: done

File: orchestra/communication/staffing.py (hunk 14 above)
Comment (kkamalov) [reply orchestra/communication/staffing.py]: all g

File: orchestra/templates/communication/tasks_available_email.txt (hunk 5 above)
Comment (kkamalov) [reply orchestra/templates/communication/tasks_available_email.txt]: g

File: orchestra/templates/communication/tasks_available_email.txt (hunk 6 above)
Comment (kkamalov) [reply orchestra/templates/communication/tasks_available_email.txt]: g

File: orchestra/templates/communication/tasks_available_slack.txt (hunk 7 above)
Comment (kkamalov) [reply orchestra/templates/communication/tasks_available_slack.txt]: d

File: orchestra/utils/task_lifecycle.py (hunk 17 above)
Comment (kkamalov) [reply orchestra/utils/task_lifecycle.py]: yep

File: orchestra/communication/staffing.py (hunk 13 above)
Comment (kkamalov) [reply orchestra/communication/staffing.py]: i'd leave it for v2

File: example_project/example_project/settings.py [hunk 18]
```diff
@@ -121,3 +124,7 @@
     PASSWORD_HASHERS = ('django.contrib.auth.hashers.MD5PasswordHasher',)
 
 setup_orchestra(__name__)
+
+
+ORCHESTRA_STAFFBOT_STAFFING_MAX_TIME = timedelta(minutes=5)
+ORCHESTRA_STAFFBOT_STAFFING_GROUP_ID = 'G7R9Q24TZ'
```
Comment (marcua) [inline example_project/example_project/settings.py]: don't include the value :)

File: orchestra/communication/utils.py [hunk 19]
```diff
@@ -0,0 +1,61 @@
+from orchestra.models import CommunicationPreference
```
Comment (marcua) [inline orchestra/communication/utils.py]: todo(marcua): read this

File: orchestra/utils/task_lifecycle.py [hunk 20]
```diff
@@ -280,6 +285,10 @@ def reassign_assignment(worker_id, assignment_id):
 
     assignment.worker = worker
     assignment.save()
+
+    mark_worker_as_winner(worker, assignment.task,
+                          assignment.assignment_counter,
```
Comment (marcua) [inline orchestra/utils/task_lifecycle.py]: sanity-checking that this is the right counter value

File: orchestra/communication/staffing.py [hunk 21]
```diff
@@ -39,8 +41,6 @@ def handle_staffing_response(worker, staffing_request_inquiry_id,
         response (orchestra.models.StaffingResponse):
             StaffingResponse object that has been created for the worker
     """
-
-    # TODO(kkamalov): add proper docstring
```
Comment (marcua) [inline orchestra/communication/staffing.py:43]: the world's longest gap from TODO to DIDNTDO :)

🌹

File: orchestra/communication/staffing.py [hunk 22]
```diff
@@ -197,3 +217,41 @@ def get_available_requests(worker):
...
+                required_role_counter=required_role_counter)
+            .order_by('-created_at'))[0]
+
+        message_experts_slack_group(
+            settings.ORCHESTRA_STAFFBOT_STAFFING_GROUP_ID,
+            ('No winner request for task {}! Created at {}'
+             .format(request.task, request.created_at)))
+
+
+def check_unstaffed_tasks():
```
Comment (marcua) [inline orchestra/communication/staffing.py]: imo: alert_workers_about_available_tasks

File: orchestra/communication/staffing.py [hunk 23]
```diff
@@ -197,3 +217,41 @@ def get_available_requests(worker):
             reverse('orchestra:communication:available_staffing_requests'))
         contexts.append(metadata)
     return contexts
+
+
+def warn_staffing_team_about_unstaffed_tasks():
```
Comment (marcua) [inline orchestra/communication/staffing.py:221]: add test :)

File: orchestra/communication/staffing.py [hunk 24]
```diff
@@ -197,3 +217,41 @@ def get_available_requests(worker):
...
+                required_role_counter=required_role_counter)
+            .order_by('-created_at'))[0]
+
+        message_experts_slack_group(
+            settings.ORCHESTRA_STAFFBOT_STAFFING_GROUP_ID,
+            ('No winner request for task {}! Created at {}.'
+             .format(request.task, request.created_at)))
+
+
+def remind_workers_about_available_tasks():
```
Comment (marcua) [inline orchestra/communication/staffing.py]: add test :)

File: orchestra/communication/staffing.py (hunk 13 above)
Comment (marcua) [reply orchestra/communication/staffing.py]: 👍

File: orchestra/communication/utils.py [hunk 25]
```diff
@@ -0,0 +1,61 @@
...
+
+    staffbot_request = staffbot_request.first()
+    if staffing_request_inquiry:
+        staffbot_request = staffing_request_inquiry.request
+
+    staffbot_request.status = StaffBotRequest.Status.COMPLETE.value
+    staffbot_request.save()
+
+    # Mark everyone else as non-winner
+    StaffingResponse.objects.filter(
```
Comment (marcua) [inline orchestra/communication/utils.py:33]: sanity-check: will this conflict with line 40 if it marks the response as not a winner?

i think not, but add that to the new/updated test :)

File: orchestra/communication/utils.py [hunk 26]
```diff
@@ -0,0 +1,61 @@
...
+from orchestra.models import StaffingRequestInquiry
+from orchestra.models import StaffingResponse
+
+import logging
+logger = logging.getLogger(__name__)
+
+
+def mark_worker_as_winner(worker, task, required_role_counter,
+                          staffing_request_inquiry):
+    staffbot_request = StaffBotRequest.objects.filter(
```
Comment (marcua) [inline orchestra/communication/utils.py]: if inquiry isn't null, should we use the request for that inquiry?

File: orchestra/communication/utils.py [hunk 27]
```diff
@@ -0,0 +1,61 @@
+from orchestra.models import CommunicationPreference
+from orchestra.models import StaffBotRequest
+from orchestra.models import StaffingRequestInquiry
+from orchestra.models import StaffingResponse
+
+import logging
+logger = logging.getLogger(__name__)
+
+
+def mark_worker_as_winner(worker, task, required_role_counter,
```
Comment (marcua) [inline orchestra/communication/utils.py]: make this transactional, or make sure everyone who calls it is transational.

(my worry is two Workers winning at the same time and both of them marking the other as not a winner :))

File: orchestra/communication/utils.py [hunk 28]
```diff
@@ -0,0 +1,61 @@
...
+                          staffing_request_inquiry):
+    staffbot_request = StaffBotRequest.objects.filter(
+        task=task, required_role_counter=required_role_counter)
```
Comment (marcua) [inline orchestra/communication/utils.py]: get the newest by created_at?

File: orchestra/communication/utils.py [hunk 29]
```diff
@@ -0,0 +1,61 @@
...
+
+    # Check whether staffbot request was sent out for this task
+    if not staffbot_request.exists():
+        return
+
+    staffbot_request = staffbot_request.first()
+    if staffing_request_inquiry:
```
Comment (marcua) [inline orchestra/communication/utils.py]: check this at the top, and put the code above in an `else` so we don't have to run queries that aren't necessary :)

File: orchestra/communication/utils.py [hunk 30]
```diff
@@ -0,0 +1,61 @@
...
+    else:
+        staffing_response = StaffingResponse.objects.filter(
+            request_inquiry__request=staffbot_request,
+            request_inquiry__communication_preference__worker=worker)
+
+    if staffing_response.exists():
+        staffing_response = staffing_response.first()
+        staffing_response.is_winner = True
+        staffing_response.save()
+    else:
```
Comment (marcua) [inline orchestra/communication/utils.py]: should we remove this else and the code in it? i'm thinking through the reason this case pops up:
- if the Worker responded to an inquiry, then they have a staffing_response and are marked as the winner above
- if the worker was assigned from projman or some other means, then there might not be an inquiry for that worker at all. so they didn't "win" and there was no "inquiry."

can we think of why this code path would happen?

Comment (marcua) [reply orchestra/communication/utils.py]: i guess the use case is if StaffBot ran, but someone assigned the task to a Worker before they received a request.

i could go either way. if you keep this code around, leave a comment explaining the stuff above :)

Comment (marcua) [reply orchestra/communication/utils.py]: now that i think of it, `warn_staffing_team_about_unstaffed_tasks` requires a winner.

so add this comment to the `else:`, assuming it's correct:

```
This branch occurs if a worker was assigned to a task that has an open StaffBot Request, but StaffBot hasn't reached out to the Worker yet. We fib a little and create in inquiry/response for the worker so they can be marked as a winner.
```

File: orchestra/communication/utils.py [hunk 31]
```diff
@@ -0,0 +1,61 @@
...
+        comm_pref = CommunicationPreference.objects.filter(
+            worker=worker)
+        if not comm_pref.exists():
+            logger.error('Worker {} does not have a communication '
+                         'preferences setup'.format(worker))
+            return
+
+        comm_pref = comm_pref.first()
+        comm_method = StaffingRequestInquiry.CommunicationMethod.SLACK.value
```
Comment (marcua) [inline orchestra/communication/utils.py:62]: read the method off the preference :)

File: orchestra/communication/utils.py [hunk 32]
```diff
@@ -0,0 +1,61 @@
...
+
+        inquiry = StaffingRequestInquiry.objects.create(
```
Comment (marcua) [inline orchestra/communication/utils.py]: only create one if `staffing_request_inquiry` is None. otherwise, use the existing inquiry

**Review by marcua — CHANGES_REQUESTED** [review]
looking good! i did a mix of live edits and leaving comments!

File: orchestra/communication/staffing.py (hunk 21 above)
Comment (kkamalov) [reply orchestra/communication/staffing.py:43]: :D

File: orchestra/communication/utils.py (hunk 27 above)
Comment (kkamalov) [reply orchestra/communication/utils.py]: sounds g

File: orchestra/communication/utils.py (hunk 28 above)
Comment (kkamalov) [reply orchestra/communication/utils.py]: sounds good

File: orchestra/communication/utils.py (hunk 29 above)
Comment (kkamalov) [reply orchestra/communication/utils.py]: coo

File: orchestra/communication/utils.py (hunk 25 above)
Comment (kkamalov) [reply orchestra/communication/utils.py:33]: i dont think so, but i will add a test

File: orchestra/communication/utils.py (hunk 31 above)
Comment (kkamalov) [reply orchestra/communication/utils.py:62]: d

File: orchestra/communication/utils.py (hunk 27 above)
Comment (kkamalov) [reply orchestra/communication/utils.py]: d

File: orchestra/communication/staffing.py [hunk 33]
```diff
@@ -256,4 +257,5 @@ def check_unstaffed_tasks():
         # TODO(kkamalov): send out reminder only if last request was sent
         # at least ORCHESTRA_STAFFBOT_MIN_FOLLOWUP_TIME ago
         if len(requests):
+            print(requests)
```
Comment (marcua) [inline orchestra/communication/staffing.py]: nope :)

File: orchestra/communication/tests/test_staffing.py [hunk 34]
```diff
@@ -366,8 +369,65 @@ def test_get_available_request(self, mock_slack):
...
+        mock_slack.assert_not_called()
+
+        create_time = timezone.now() - timedelta(minutes=31)
+        StaffingResponseFactory(
+            request_inquiry__request__task__start_datetime=create_time,
+            request_inquiry__request__created_at=create_time,
+            is_winner=False
+        )
+        warn_staffing_team_about_unstaffed_tasks()
+        self.assertEqual(mock_slack.call_count, 1)
```
Comment (marcua) [inline orchestra/communication/tests/test_staffing.py:386]: make sure it's the right message :)

File: orchestra/communication/tests/test_staffing.py [hunk 35]
```diff
@@ -366,8 +369,65 @@ def test_get_available_request(self, mock_slack):
...
+            is_available=True,
+            is_winner=True)
+
+        check_unstaffed_tasks()
+        mock_staffbot.assert_not_called()
+
+        staffing_response.delete()
+
+        check_unstaffed_tasks()
+        self.assertEqual(mock_staffbot.call_count, 1)
```
Comment (marcua) [inline orchestra/communication/tests/test_staffing.py]: make sure it's the right message :)

File: orchestra/communication/utils.py [hunk 36]
```diff
@@ -50,12 +61,14 @@ def mark_worker_as_winner(worker, task, required_role_counter,
         comm_pref = comm_pref.first()
         comm_method = StaffingRequestInquiry.CommunicationMethod.SLACK.value
```
Comment (marcua) [inline orchestra/communication/utils.py]: read off preference

File: orchestra/communication/tests/test_staffing.py [hunk 37]
```diff
@@ -3,32 +3,48 @@
 from unittest.mock import patch
 
 from orchestra.bots.errors import StaffingResponseException
+from orchestra.communication.staffing import check_unstaffed_tasks
```
Comment (marcua) [inline orchestra/communication/tests/test_staffing.py]: name change after merge

File: orchestra/communication/staffing.py (hunk 23 above)
Comment (kkamalov) [reply orchestra/communication/staffing.py:221]: done

File: orchestra/communication/tests/test_staffing.py (hunk 37 above)
Comment (kkamalov) [reply orchestra/communication/tests/test_staffing.py]: done

File: orchestra/communication/tests/test_staffing.py (hunk 34 above)
Comment (kkamalov) [reply orchestra/communication/tests/test_staffing.py:386]: done

**Review by marcua — APPROVED** [review]
Looks good assuming:
1) you manually tested,
2) you fix the `comm_method` comment :)

**Pull request closed without merging**